node/subsystem/src/messages.rs

```diff
@@ -208,6 +209,73 @@ pub enum CollatorProtocolMessage {
 }
 
+/// Messages received by the dispute coordinator subsystem.
+#[derive(Debug)]
+pub enum DisputeCoordinatorMessage {
+	/// Import a statement by a validator about a candidate.
+	///
+	/// The subsystem will silently discard ancient statements or sets of only dispute-specific statements for
+	/// candidates that are previously unknown to the subsystem. The former is simply because ancient
+	/// data is not relevant and the latter is as a DoS prevention mechanism. Both backing and approval
+	/// statements already undergo anti-DoS procedures in their respective subsystems, but statements
+	/// cast specifically for disputes are not necessarily relevant to any candidate the system is
+	/// already aware of and thus present a DoS vector. Our expectation is that nodes will notify each
+	/// other of disputes over the network by providing (at least) 2 conflicting statements, of which one is either
+	/// a backing or validation statement.
+	///
+	/// This does not do any checking of the message signature.
+	ImportStatements {
+		/// The hash of the candidate.
+		candidate_hash: CandidateHash,
+		/// The candidate receipt itself.
+		candidate_receipt: CandidateReceipt,
+		/// The session the candidate appears in.
+		session: SessionIndex,
+		/// Triples containing the following:
+		/// - A statement, either indicating validity or invalidity of the candidate.
+		/// - The validator index (within the session of the candidate) of the validator casting the vote.
+		/// - The signature of the validator casting the vote.
+		statements: Vec<(DisputeStatement, ValidatorIndex, ValidatorSignature)>,
+	},
+	/// Fetch a list of all active disputes that the co-ordinator is aware of.
+	ActiveDisputes(oneshot::Sender<Vec<(SessionIndex, CandidateHash)>>),
+	/// Get candidate votes for a candidate.
+	QueryCandidateVotes(SessionIndex, CandidateHash, oneshot::Sender<Option<CandidateVotes>>),
+	/// Sign and issue local dispute votes. A value of `true` indicates validity, and `false` invalidity.
+	IssueLocalStatement(SessionIndex, CandidateHash, CandidateReceipt, bool),
+	/// Determine the highest undisputed block within the given chain, based on where candidates
+	/// were included. If even the base block should not be finalized due to a dispute,
+	/// then `None` should be returned on the channel.
+	///
+	/// The block descriptions begin counting upwards from the block after the given `base_number`. The `base_number`
+	/// is typically the number of the last finalized block but may be slightly higher. This block
+	/// is inevitably going to be finalized so it is not accounted for by this function.
+	DetermineUndisputedChain {
+		/// The number of the lowest possible block to vote on.
+		base_number: BlockNumber,
+		/// Descriptions of all the blocks counting upwards from the block after the base number
+		block_descriptions: Vec<(Hash, SessionIndex, Vec<CandidateHash>)>,
+		/// A response channel - `None` to vote on base, `Some` to vote higher.
```

Comment:
Isn't `None` meant to mean that not even the base is valid? I.e. if we want to vote on base we should return `Some((base_number, base_hash))`. I suppose `base` will already be finalized though.

Reply:
Related: https://github.com/paritytech/polkadot/pull/3133#issuecomment-850602345.

Reply:
My understanding of the GRANDPA logic is that base is already 'inescapable'. The minimal possible vote is on the base